packages/jaeger-ui/src/utils/link-formatting.tsx

```diff
@@ -49,31 +49,66 @@ function getFormatFunctions<T = Trace[keyof Trace]>(): Record<
 
       return value.padStart(desiredLength, padCharacter);
     },
+
+    add: (value, offsetString: string) => {
+      if (typeof value !== 'number') {
+        console.error('add can only operate on numbers, ignoring formatting', {
+          value,
+          offsetString,
+        });
+        return value;
+      }
+
+      const offset = parseInt(offsetString, 10);
+      if (Number.isNaN(offset)) {
+        console.error('add needs a valid offset in microseconds as second argument, ignoring formatting', {
+          value,
+          offsetString,
+        });
+        return value;
+      }
+
+      return value + offset;
+    },
   };
 }
 
 export function getParameterAndFormatter<T = Trace[keyof Trace]>(
   parameter: string
 ): {
   parameterName: string;
-  formatFunction: ((value: T) => T | string) | null;
+  formatFunction: ((value: T) => T | string | number) | null;
 } {
   const parts = parameter.split('|').map(part => part.trim());
   const parameterName = parts[0];
   if (parts.length === 1) return { parameterName, formatFunction: null };
 
-  const [formatFunctionName, ...args] = parts[1].split(' ');
-
   const formatFunctions = getFormatFunctions<T>();
 
-  const formatFunction = formatFunctions[formatFunctionName];
-  if (!formatFunction) {
-    console.error('Unrecognized format function name, ignoring formatting', {
-      parameter,
-      formatFunctionName,
-      validValues: Object.keys(formatFunctions),
-    });
-  }
+  const chainedFormatFunctions = parts
+    .slice(1)
+    .map(formatPart => {
+      const [formatFunctionName, ...args] = formatPart.split(' ');
+      const formatFunction = formatFunctions[formatFunctionName];
+      if (!formatFunction) {
+        console.error(
+          'Unrecognized format function name, ignoring formatting. Other formatting functions may be applied',
+          {
+            parameter,
+            formatFunctionName,
+            validValues: Object.keys(formatFunctions),
+          }
+        );
+        return null;
+      }
+      return (val: T) => formatFunction(val, ...args);
+    })
+    .filter((fn): fn is NonNullable<typeof fn> => fn !== null);
```

Comment:
odd statement, NonNullable is compared against null?

I think you could use flatMap() instead and return `[]` in cases of errors such that it will be automatically skipped

Reply:
This is a [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates) which says that it returns a boolean indicating if `fn` is non-null. It may make more sense reading it with a curly brace-wrapped function body
```typescript
.filter((fn): fn is NonNullable<typeof fn> => {
  return fn !== null;
});
```

I could use flatMap, but would just add more allocations for the same behavior right?